File: evaluateDivision/evaluateDivision.java
class Solution {
    public double[] calcEquation(List<List<String>> equations, double[] values, List<List<String>> queries) {
        Map<String, Map<String, Double>> graph = new HashMap<>();
        for (int i = 0; i < equations.size(); i++) {
            List<String> equation = equations.get(i);
            if (!graph.containsKey(equation.get(0))) {
                graph.put(equation.get(0), new HashMap<>());
            }
            graph.get(equation.get(0)).put(equation.get(1), values[i]);
            if (!graph.containsKey(equation.get(1))) {
                graph.put(equation.get(1), new HashMap<>());
            }
            graph.get(equation.get(1)).put(equation.get(0), 1 / values[i]);
        }
        double[] res = new double[queries.size()];
        for (int i = 0; i < queries.size(); i++) {
            if (!graph.containsKey(queries.get(i).get(0)) || !graph.containsKey(queries.get(i).get(1))) {
                res[i] = -1.0;
            } else {
                res[i] = dfs(graph, queries.get(i).get(0), queries.get(i).get(1), new HashSet<>());
            }
        }
        return res;
    }
    
    public double dfs(Map<String, Map<String, Double>> graph, String start, String end, Set<String> visited) {
        if (start.equals(end)) return 1;
        
        for (String neighbor : graph.get(start).keySet()) {
            if (!visited.contains(neighbor)) {
                visited.add(neighbor);
                double neighorRes = dfs(graph, neighbor, end, visited);
                if (neighorRes != -1.0) {
                    return graph.get(start).get(neighbor) * neighorRes;
                }
                visited.remove(neighbor);
            }
        }
        return -1;
    }
}

File: 399.evaluateDivision/evaluateDivision.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <unordered_set>
using namespace std;

class Solution {
public:
    vector<double> calcEquation(vector<pair<string, string>> equations, vector<double>& values, vector<pair<string, string>> queries) {
        unordered_map<string, vector<string> > edges;
        unordered_map<string, vector<double> > edgesValue;
        for (int i = 0; i < equations.size(); i++) {
            edges[equations[i].first].push_back(equations[i].second);
            edgesValue[equations[i].first].push_back(values[i]);
            edges[equations[i].second].push_back(equations[i].first);
            edgesValue[equations[i].second].push_back(1 / values[i]);
        }
        vector<double> res(queries.size(), 0.0);
        for (int i = 0; i < queries.size(); i++) {
            unordered_set<string> visited;
            double curValue = 1.0;
            res[i] = dfs(queries[i].first, queries[i].second, visited, edges, edgesValue, curValue);
            if (res[i] == 0.0)
                res[i] = -1.0;
        }
        return res;
    }
    double dfs(string start, string end, unordered_set<string>& visited, unordered_map<string, vector<string> >& edges, unordered_map<string, vector<double> >& edgesValue, double curValue) {
        if (!edges.count(start))
            return 0.0;
        if (visited.count(start))
            return 0.0;
        if (start == end)
            return curValue;
        visited.insert(start);
        for (int i = 0; i < edges[start].size(); i++) {
            double tmp = dfs(edges[start][i], end, visited, edges, edgesValue, curValue * edgesValue[start][i]);
            if (tmp != 0.0)
                return tmp;
        }
        visited.erase(start);
        return 0.0;
    }
};

int main() {
    Solution s;
    vector<pair<string, string> > equations = {make_pair("a", "b"), make_pair("b", "c")};
    vector<double> values = {2.0, 3.0};
    vector<pair<string, string> > queries = {make_pair("a", "c"), make_pair("b", "a"), make_pair("a", "e"), make_pair("a", "a"),make_pair("x", "x")};
    for (double& d : s.calcEquation(equations, values, queries)) {
        cout << d << " ";
    }
    cout << endl;
    return 0;
}
